sstables/generation_type.cc

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "sstables/generation_type.hh"
+#include "api/api.hh"
+#include <cctype>
+#include <fmt/core.h>
+#include <cstdint>
+#include <iterator>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace {
+sstring to_base36(uint64_t i) {
+    std::vector<char> buf(13, '\0');
+    auto s = buf.begin();
+    do {
+        *s = "0123456789abcdefghijklmnopqrstuvwxyz"[i % 36];
+        i /= 36;
+        s++;
+    } while (i);
+
+    return sstring(std::make_reverse_iterator(s), buf.rend());
+}
+
+uint64_t from_base36(const sstring& buf) {
+    uint64_t total = 0;
+    for (auto i = buf.size() - 1; i >= 0; --i) {
+        total *= 36;
+        const char c = std::tolower(buf[i]);
+        total += (c >= 'a' && c <= 'z') ? c - 'a' + 10 : c - '0';
+    }
+    return total;
+}
+} // anon. namespace
+
+namespace seastar {
+sstring to_sstring(const sstables::generation_type& gen) {
+    return std::visit(make_visitor(
+        [] (int64_t x) { return fmt::format("{}", x); },
+        [] (utils::UUID u) { 
+            auto ts = u.timestamp();
+            const auto nano_part = ts % 10'000'000;
+            ts /= 10'000'000;
+            const auto seconds = ts % 86'400;
+            ts /= 86'400;
+            return fmt::format("{:0>4s}_{:0>4s}_{:0>5s}{:0>13s}",
+                to_base36(ts),
+                to_base36(seconds),
+                to_base36(nano_part),
+                to_base36(u.get_least_significant_bits()));
+        }
+    ), gen.value);
+}
+} // namespace seastar
+
+namespace sstables {
+int calc_shard_num(const generation_type& gen) {
+    return std::visit(make_visitor(
+        [] (int64_t x) { return int(x % smp::count); },
+        [] (utils::UUID u) -> int { throw std::runtime_error("unimplemented"); }
+    ), gen.value);
+}
```

Comment:
Is it used anywhere?

Reply:
I'm not sure where you want this to be used.

Note that we currently don't use the generation number to infer the owner shard, but rather we use the metadata provided by `get_shards_for_this_sstable` to determine which shard should server the sstable, so it could work wither with either integer or uuid-type  generation.

See https://github.com/scylladb/scylladb/blob/a7c4a129cb349a51a0df870db37a26eb29f7d20e/sstables/sstable_directory.cc#L139-L143

Reply:
Previously we used generation % sharad as a hint for owner, as sstable could be shared by multiple shards. If we still do it, it's unnecessary and can be removed.

Reply:
I don't think we still use that.
@jul-stas did you encounter a use site where we defer the sstable shard owner from its generation?

Reply:
```
[raphaelsc@localhost scylla]$ grep -r 'calculate_shard_from_sstable_generation' --include=*.cc
test/boost/sstable_resharding_test.cc:        BOOST_REQUIRE(column_family_test::calculate_shard_from_sstable_generation(generation_value(new_sst->generation())) == shard);
```


we can just kill that.


this is enough:
```
        auto shards = new_sst->get_shards_for_this_sstable();
        BOOST_REQUIRE(shards.size() == 1); // check sstable is unshared.
```

Reply:
Thanks.  I see you submitted https://github.com/scylladb/scylladb/pull/12476